pkg/server/plugin/upstreamauthority/spire/spire_workload.go

```diff
@@ -5,32 +5,59 @@ import (
 	"errors"
 	"net"
 
+	"github.com/hashicorp/go-hclog"
 	proto "github.com/spiffe/go-spiffe/proto/spiffe/workload"
 	"github.com/spiffe/spire/api/workload"
+	"github.com/spiffe/spire/proto/spire/api/node"
 )
 
-func (m *Plugin) getWorkloadSVID(ctx context.Context, config *Configuration) ([]byte, []byte, []byte, error) {
+func (m *Plugin) watchWorkloadSVID(ctx context.Context, workloadAPISocket string, ready chan struct{}) {
 	errorChan := make(chan error, 1)
-
-	wapiClient := m.newWorkloadAPIClient(config.WorkloadAPISocket)
+	wapiClient := m.newWorkloadAPIClient(workloadAPISocket)
 	updateChan := wapiClient.UpdateChan()
 	go func() {
 		err := wapiClient.Start()
 		if err != nil {
 			errorChan <- err
 		}
 	}()
-
 	defer wapiClient.Stop()
 
+	firstDone := false
 	for {
 		select {
 		case svidResponse := <-updateChan:
-			return m.receiveUpdatedCerts(svidResponse)
+			wCert, wKey, wBundle, err := m.receiveUpdatedCerts(svidResponse)
+			if err != nil {
+				m.log.Error("cannot receive workload SVID update", "error", err)
+				continue
+			}
+
+			conn, err := m.newNodeClientConn(ctx, wCert, wKey, wBundle)
+			if err != nil {
+				m.log.Error("cannot create node client gRPC connection", "error", err)
+				continue
+			}
+
+			m.nodeMtx.Lock()
+			if m.conn != nil {
+				m.conn.Close()
+			}
+			m.conn = conn
+			m.nodeClient = node.NewNodeClient(m.conn)
+			m.nodeMtx.Unlock()
+
+			if !firstDone {
+				firstDone = true
+				close(ready)
+			}
+
 		case <-ctx.Done():
-			return nil, nil, nil, ctx.Err()
+			m.log.Debug("watch workload context done", "reason", ctx.Err())
```

Comment:
Should the `nodeClient` and `conn` be cleaned up before returning here?

Reply:
Right!